File: problemset1/credit/credit.c
#include <stdio.h>
#include <math.h>

int company(long long card, int length);
int luhn(long long card, int length);

int main(){
    long long card;
    
    printf("Enter your credit card number here: ");
    scanf("%lli", &card);;

    int length = log10(card);
    length++;

    int valid = luhn(card, length);

    if(valid == 1){
        int comp = company(card, length);
        if(comp == 1){
            printf("VISA");
        }
        else if(comp == 2){
            printf("MASTERCARD");
        }
        else if(comp == 3){
            printf("AMERICAN EXPRESS");
        }
        else{
            printf("INVALID LENGTH");
        }
    }
    else{
        printf("INVALID LUHN EQUATION");
    }
    return 0;
}

int company(long long card, int length){
    int first = (int)(card / pow(10, length - 1));

    int mastercard[] = {51, 52, 53, 54, 55};
    int americanexpress[] = {34, 37};
    int visa = 4;

    if(first == visa){
        if(length == 13 || length == 16){
            return 1;
        }
    }

    int first2 = (int)(card / pow(10, length-2));

    for(int i = 0; i < 5; i++){
        if(first2 == mastercard[i]){
            if(length == 16){
                return 2;
            }
            return 0;
        }
    }
    for(int i = 0; i < 2; i++){
        if(first2 == americanexpress[i]){
            if(length == 15){
                return 3;
            }
            return 0;
        }
    }
    return 0;
}

int luhn(long long card, int length){
    int array[length];
    long long tempcard = card;
    int test = 0;

    for(int i = 0; i < length; i++){
        array[length-(i+1)] = tempcard % 10;
        tempcard /= 10;
    }

    int sizeOfDoubled;
    int odd = 0;
    if(length%2 == 0){
        sizeOfDoubled = length/2;
    }
    else{
        sizeOfDoubled = (length-1)/2;
        odd = 1;
    }

    int doubled[sizeOfDoubled];
    int j = 0;
    for(int i = length-2; i > 0; i-=2){
        doubled[j] = array[i] * 2;
        j++;
    }

    if(odd == 1){
        for(int i = 0; i < sizeOfDoubled; i++){
            if(doubled[i] > 9){
                test += doubled[i] % 10;
                doubled[i] /= 10;
                test += doubled[i];
            }
            else{
                test += doubled[i];
            }
        }
    }
    else{
        for(int i = 0; i < sizeOfDoubled-1; i++){
            if(doubled[i] > 9){
                test += doubled[i] % 10;
                doubled[i] /= 10;
                test += doubled[i];
            }
            else{
                test += doubled[i];
            }
        }
        int x = array[0] * 2;
        if(x > 9){
            test += x % 10;
            x /= 10;
            test += x;
        }
        else{
            test += array[0] * 2;
        }
    }

    for(int i = length-1; i >= 0; i-=2){
        test += array[i];
    }

    if(test % 10 == 0){
        return 1;
    }
    return 0;
}

File: problemset6/sentimental-credit/credit.py
from sys import exit

def integerCheck(card):
    try:
        int(card)
        return True
    except ValueError:
        return False

def luhn(card, length):
    array = []
    for i in range(length):
        array.append(card[i])
        array[i] = int(array[i])
        
    doubled = []
    counter = length - 2
    while counter >= 0:
        doubled.append(array[counter] * 2)
        counter -= 2
            
    for i in range(len(doubled)):
        doubled[i] = doubled[i]
        if doubled[i] >= 10:
            num = str(doubled[i])
            for j in range(len(num)):
                doubled.append(int(num[j]))
    
    doubled.sort()
    new = []
    
    for i in range(len(doubled)):
        if doubled[i] < 10:
            new.append(doubled[i])
            
                
    num = 0
    for i in range(len(new)):
        num += new[i]
    
    counter2 = length - 1
    while counter2 >= 0:
        num += array[counter2]
        counter2 -= 2
    
    if num % 10 == 0:
        return True
    return False

def company(card, length):
    mastercard = [51, 52, 53, 54, 55]
    amex = [34, 37]
    visa = [4]
    
    if int(card[0]) in visa:
        if length == 13 or length == 16:
            return "visa"
    
    elif int(card[0] + card[1]) in mastercard:
        if length == 16:
            return "mastercard"
    
    elif int(card[0] + card[1]) in amex:
        if length == 15:
            return "amex"
    
    return "invalid"

def main():
    card = input("Enter card number: ")
    if integerCheck(card) == False:
        print("INVALID")
        exit(1)
        
    length = len(card)

    valid = luhn(card, length)

    if valid == True:
        comp = company(card, length)
        if comp == "visa":
            print("VISA")
        elif comp == "mastercard":
            print("MASTERCARD")
        elif comp == "amex":
            print("AMERICAN EXPRESS")
        else:
            print("INVALID")

    else:
        print("INVALID")
        
main()